torchvision/prototype/datasets/utils/_decoder.py

```diff
@@ -0,0 +1,47 @@
+from typing import Any, Dict, Callable
+from typing import BinaryIO
+
+import PIL.Image
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import Mapper, IterDataPipe
+from torchvision.prototype import features
+from torchvision.transforms.functional import pil_to_tensor
+
+
+def decode_image_with_pil(buffer: BinaryIO) -> Dict[str, Any]:
+    return dict(image=features.Image(pil_to_tensor(PIL.Image.open(buffer))))
+
+
+class DecodeableStreamWrapper:
+    def __init__(self, stream: BinaryIO, decoder: Callable[[BinaryIO], Dict[str, Any]]) -> None:
+        self.__stream__ = stream
+        self.__decoder__ = decoder
+
+    # TODO: dispatch attribute access besides `decode` to `__stream__`
+
+    def decode(self) -> Dict[str, Any]:
+        return self.__decoder__(self.__stream__)
+
+    def unwrap(self) -> BinaryIO:
+        return self.__stream__
+
+
+class DecodeableImageStreamWrapper(DecodeableStreamWrapper):
+    def __init__(self, stream: BinaryIO, decoder: Callable[[BinaryIO], Dict[str, Any]] = decode_image_with_pil) -> None:
+        super().__init__(stream, decoder)
+
+
+def decode_sample(sample: Dict[str, Any]) -> Dict[str, Any]:
+    decoded_sample = dict()
+    for name, obj in sample.items():
+        if isinstance(obj, DecodeableStreamWrapper):
+            decoded_sample.update(obj.decode())
+        else:
+            decoded_sample[name] = obj
+    return decoded_sample
+
+
+@functional_datapipe("decode_samples")
+class SampleDecoder(Mapper[Dict[str, Any]]):
+    def __init__(self, datapipe: IterDataPipe[Dict[str, Any]]) -> None:
+        super().__init__(datapipe, decode_sample)
```

Comment:
So, how can user provide a custom decoder? Then, do we want users add their own `StreamWrapper` at the end of  each pipeline.

Reply:
The idea of the design is that each `DecodableStreamWrapper` dispatches all attributes to wrapped file object (this is not yet implemented). With that there is no need to provide a mechanism for custom decoders, because they can just use the file directly. Imagine I want to decode an image with my custom decoder, while using the default one for the annotations:

```python
from torchvision.prototype import datasets

def custom_image_decoder(buffer: BinaryIO) -> Dict[str, Any]:
    ...

def custom_decode_sample(sample):
    sample.update(custom_image_decoder(sample.pop("image")))
    return datasets.utils.decode_sample(sample)


dataset = datasets.load(...).map(custom_decode_sample)
```

Reply:
But, then where this `decode_samples` DataPipe API can be used with custom decoder?

Reply:
See https://github.com/pytorch/vision/pull/5105#issuecomment-999673849